### PR TITLE
Added density of a categorical #10949

### LINKED ISSUE #10949: ENH: density of a categorical
I think ``s.cat.density`` is appropriate here

```
In [1]: s = Series(list('aabbcccc')).astype('category')

In [2]: 100*len(s.cat.categories)/len(s)
Out[2]: 37
```

### PR DESCRIPTION
**Pull request by sornars**
Not sure if this is the correct approach to the problem 

closes #10949.

 Assuming it is I will go ahead and add some documentation for this enhancement. Let me know if you have any feedback.

### CODE REVIEW

**Comment by maximilianr** [comment]
Great!

I know @jreback suggested to multiply by 100, but is that right? Shouldn't we let the user convert to percent? (a more general point than this PR)

I'm not sure you want integer division there.

I would test for an actual value, rather than just the code that you wrote

**Comment by sornars** [comment]
Happy to change this to a proportion, I agree that it seems a more sensible default. 

With regards to integer division, what is the preferred way to target Python 2 and 3 with float division, `from __future__ import division`?

Will update the tests in the next commit.

**Comment by maximilianr** [comment]
I'm not sure on the pandas standard for the division - the maintainers will opine - either `__future__` or coercing with `float(x)` would work.

And they'll also ask for a note in what's new & the docs.

Cheers!

File: pandas/core/categorical.py [hunk 1]
```diff
@@ -1753,6 +1754,10 @@ def codes(self):
         from pandas import Series
         return Series(self.categorical.codes, index=self.index)
 
+    @property
+    def density(self):
+        return len(self.categorical.categories)/len(self.categorical)
```
Comment (jreback) [inline pandas/core/categorical.py]: yeh just do ``float(len(self.categorical))`` to force truediv

Comment (jreback) [inline pandas/core/categorical.py]: pls add a doc-string (and add in API.rst where the rest of methods are)

**Comment by jreback** [comment]
further, can you do a tiny bit of research to see what ``R`` calls this (they call the ``Categorical`` a ``Factor``.

**Comment by sornars** [comment]
I haven't had much luck searching for an equivalent field in R but I have found that the SAS community calls this number the cardinality ratio: http://www.sascommunity.org/wiki/Cardinality_Ratio

**Comment by kawochen** [comment]
Should we consider preempting or handling `ZeroDivisionError`?

**Comment by jreback** [comment]
whoosh never realized we allow 0 categories!. Yes that should be a ``NaN``

File: pandas/core/categorical.py [hunk 2]
```diff
@@ -1753,6 +1753,14 @@ def codes(self):
         from pandas import Series
         return Series(self.categorical.codes, index=self.index)
 
+    @property
+    def density(self):
+        """The cardinality ratio of this categorical."""
+        if len(self.categorical):
+            return float(len(self.categorical.categories))/len(self.categorical)
+        else:
```
Comment (jreback) [inline pandas/core/categorical.py]: you can just return here (don't need else)

File: pandas/tests/test_categorical.py [hunk 3]
```diff
@@ -3651,6 +3651,12 @@ def test_concat_categorical(self):
 
         tm.assert_frame_equal(df_expected, df_concat)
 
+    def test_density(self):
+        s1 = Series(list('aabbcccc')).astype('category')
+        self.assertEqual(0.375, s1.cat.density)
+
+        s2 = Series([]).astype('category')
+        tm.assert_numpy_array_equal(np.nan, s2.cat.density)
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: ``self.assertEqual(np.nan, s2.cat.density)``

**Comment by jreback** [comment]
pls add a release note in 0.17.1 (enhancements)
add to api.rst as well

**Comment by jreback** [comment]
can you update

File: pandas/tests/test_categorical.py (hunk 3 above)
Comment (sornars) [inline pandas/tests/test_categorical.py]: Quick question on this change: np.nan does not equal np.nan so this test will fail. Am I correct in thinking that s2.cat.density should equal np.nan?

**Comment by sornars** [comment]
Apologies for the delay, I was trying to understand out the correct way to add documentation for this feature before other things took ahold of me. 

I've made an updated commit. There remains one failing test due to the recommendation to switch from `tm.assert_numpy_array_equal(np.nan, s2.cat.density)` to `self.assertEqual(np.nan, s2.cat.density)`. Please advise on the best way to approach this failing test.

**Comment by jreback** [comment]
use self.assertTrue(... is np.nan)

File: pandas/core/categorical.py [hunk 4]
```diff
@@ -1753,6 +1753,11 @@ def codes(self):
         from pandas import Series
         return Series(self.categorical.codes, index=self.index)
 
+    @property
+    def density(self):
+        """The cardinality ratio of this categorical."""
+        return np.divide(float(len(self.categorical.categories)), len(self.categorical))
```
Comment (kawochen) [inline pandas/core/categorical.py]:
```
s1 = Series(list('aabbcccc')).astype('category')
s2 = s1[1:1]
```
Then `s2.cat.density` would return `inf`?

Comment (sornars) [inline pandas/core/categorical.py]: What would be a more reasonable return value in this instance? `nan`? I think an `inf` value suggests that `k >> n` which could be useful (ie. categories are defined but the categorical is empty) as opposed to `nan` which implies that ratio is meaningless: `k = n = 0`.

**Comment by jreback** [comment]
closing as we closed the original issue #10949 

let's see an actual usecase (as I proposed the original issue, and don't really have one :)

**Pull request closed without merging**